### PR TITLE
feat(auth): add universe domain to grpctransport and httptransport

### PR DESCRIPTION
**Pull request by quartzmo**
This pull request introduces an intentional breaking change by removing `httptransport.InternalOptions.DefaultEndpoint` and `grpctransport.InternalOptions.DefaultEndpoint`. This is because [internaloption.WithDefaultEndpoint](https://pkg.go.dev/google.golang.org/api@v0.172.0/option/internaloption#WithDefaultEndpoint) is now deprecated, and new code should instead use the combination of [internaloption.WithDefaultEndpointTemplate](https://pkg.go.dev/google.golang.org/api@v0.172.0/option/internaloption#WithDefaultEndpointTemplate) and [internaloption.WithDefaultUniverseDomain](https://pkg.go.dev/google.golang.org/api@v0.172.0/option/internaloption#WithDefaultUniverseDomain).

Adapted from: https://github.com/googleapis/google-api-go-client/pull/2296

Fixes: #9670

### CODE REVIEW

File: auth/credentials/impersonate/impersonate_test.go [hunk 1]
```diff
@@ -25,38 +25,54 @@ import (
 	"time"
 
 	"github.com/google/go-cmp/cmp"
+	"google.golang.org/api/option"
```
Comment (codyoss) [inline auth/credentials/impersonate/impersonate_test.go:28]: This module intentionally does not rely on `google.golang.org/api`. This does not look needed for the tests.

File: auth/go.mod [hunk 2]
```diff
@@ -9,6 +9,7 @@ require (
 	github.com/googleapis/enterprise-certificate-proxy v0.3.2
 	go.opencensus.io v0.24.0
 	golang.org/x/net v0.22.0
+	google.golang.org/api v0.149.0
```
Comment (codyoss) [inline auth/go.mod:12]: `go mo tidy` after removal

File: auth/grpctransport/grpctransport.go [hunk 3]
```diff
@@ -143,13 +143,17 @@ type InternalOptions struct {
 	// DefaultAudience specifies a default audience to be used as the audience
 	// field ("aud") for the JWT token authentication.
 	DefaultAudience string
-	// DefaultEndpoint specifies the default endpoint.
-	DefaultEndpoint string
+	// DefaultEndpointTemplate combined with DefaultUniverseDomain specifies
+	// the default endpoint.
+	DefaultEndpointTemplate string
```
Comment (codyoss) [inline auth/grpctransport/grpctransport.go:148]: Will this also be done for mTLS in the future? If so let's rename it now as well here so we are not stuck in the future. Rather make all breaking changes at once.

File: auth/internal/transport/cba.go [hunk 4]
```diff
@@ -40,21 +42,90 @@ const (
...
+		return ""
+	}
+	return strings.Replace(o.DefaultEndpointTemplate, universeDomainPlaceholder, o.getUniverseDomain(), 1)
+}
+
+// mergedEndpoint merges a user-provided Endpoint of format host[:port] with the
+// default endpoint.
+func (o *Options) mergedEndpoint() (string, error) {
+	defaultEndpoint := o.defaultEndpoint()
+	if defaultEndpoint == "" {
```
Comment (codyoss) [inline auth/internal/transport/cba.go:108]: This is not a field that users set. It should be valid to not have any InternalOption sent.

File: auth/internal/transport/s2a_test.go [hunk 5]
```diff
@@ -93,3 +94,37 @@ func TestMTLSConfigExpiry(t *testing.T) {
...
+		{
+			name:      "empty",
+			useS2AEnv: "",
+			want:      false,
+		},
+	}
+
+	for _, tc := range testCases {
+		if tc.useS2AEnv != "" {
+			os.Setenv(googleAPIUseS2AEnv, tc.useS2AEnv)
```
Comment (codyoss) [inline auth/internal/transport/s2a_test.go:123]: You don't want to use os.Setenv is tests as it changes the state for the whole test run. Instead you can use: https://pkg.go.dev/testing#T.Setenv which restores the value after.

**Review by codyoss — CHANGES_REQUESTED** [review]

File: auth/internal/transport/s2a_test.go (hunk 5 above)
Comment (quartzmo) [reply auth/internal/transport/s2a_test.go:123]: Yes, I meant to use `t.Setenv`, thanks for catching this.

File: auth/credentials/impersonate/impersonate_test.go (hunk 1 above)
Comment (quartzmo) [reply auth/credentials/impersonate/impersonate_test.go:28]: Fixed, thanks!

File: auth/go.mod (hunk 2 above)
Comment (quartzmo) [reply auth/go.mod:12]: Fixed, thanks!

File: auth/grpctransport/grpctransport.go (hunk 3 above)
Comment (quartzmo) [reply auth/grpctransport/grpctransport.go:148]: Gonna leave this for now, per offline discussion.

File: auth/grpctransport/grpctransport.go [hunk 6]
```diff
@@ -143,13 +143,17 @@ type InternalOptions struct {
...
 	// DefaultMTLSEndpoint specifies the default mTLS endpoint.
 	DefaultMTLSEndpoint string
 	// DefaultScopes specifies the default OAuth2 scopes to be used for a
 	// service.
 	DefaultScopes []string
+	// DefaultUniverseDomain is the default value for universe domain.
+	// Universe domain is the default service domain for a given Cloud universe.
+	DefaultUniverseDomain string
```
[the remark this answers is not in the document]
Comment (quartzmo) [reply auth/grpctransport/grpctransport.go]: Removed from here and httptransport, thanks.

File: auth/internal/transport/cba.go (hunk 4 above)
Comment (quartzmo) [reply auth/internal/transport/cba.go:95]: Removed, thanks.

File: auth/internal/transport/cba_test.go [hunk 7]
```diff
@@ -383,3 +475,181 @@ func setupTest(t *testing.T) func() {
...
+				ClientCertProvider:      fakeClientCertSource,
+			},
+			wantEndpoint: testUniverseDomainEndpoint,
+			wantErr:      errUniverseNotSupportedMTLS,
+		},
+	}
+
+	for _, tc := range testCases {
+		if tc.opts.ClientCertProvider != nil {
+			os.Setenv("GOOGLE_API_USE_CLIENT_CERTIFICATE", "true")
```
[the remark this answers is not in the document]
Comment (quartzmo) [reply auth/internal/transport/cba_test.go]: Fixed.

File: auth/internal/transport/cba_test.go [hunk 8]
```diff
@@ -58,49 +64,132 @@ var (
...
+			wantDefaultUniverseDomain: "googleapis.com",
+			wantUniverseDomain:        "googleapis.com",
+			wantDefaultEndpoint:       "https://test.googleapis.com/bar/baz",
+			wantIsGDU:                 true,
+			wantMergedEndpoint:        "https://myhost:8000/bar/baz",
+		},
+	}
+	for _, tc := range testCases {
+		if got := tc.opts.getDefaultUniverseDomain(); got != tc.wantDefaultUniverseDomain {
+			t.Errorf("%s: got %v, want %v", tc.name, got, tc.wantDefaultUniverseDomain)
```
[the remark this answers is not in the document]
Comment (quartzmo) [reply auth/internal/transport/cba_test.go]: Done, thanks!

**Pull request merged**